### DISCUSSION TITLE
p5 Versioning

### DISCUSSION THREAD

**Discussion by taliacotton** [P5-wrapper/react#317] [Q&A] [answered]
I've noticed the component depends on a specific version of p5 (as of right now I see 1.9.0). I was wondering if there's a way to control which version of p5 it uses. The reason for this is that there are certain features of p5 in prior versions, so I'm wondering if it's possible to downgrade the version. Thanks!

```
// package-lock.json
    "node_modules/@p5-wrapper/react": {
      "version": "4.3.2",
      "resolved": "https://registry.npmjs.org/@p5-wrapper/react/-/react-4.3.2.tgz",
      "integrity": "sha512-m7MoL63grVVyu39czEi3V3tz29VM4zzd4K9MjNWZ5t7jb4itgHpw6FCiSG2JzU1SqElKgo9ny8NliavPWYYdrw==",
      "dependencies": {
        "microdiff": "^1.3.2",
        "p5": "^1.9.0"
      },
      "peerDependencies": {
        "react": ">= 18.2.0",
        "react-dom": ">= 18.2.0"
      }
    },
```

**Comment by yevdyko** [comment]
@taliacotton I guess you can try to install a particular version of p5 explicitly in your project dependencies

**Reply by taliacotton** [reply]
Yep, tried that! Here's what I've tried so far:

```
npm uninstall p5
npm uninstall @p5-wrapper/next
npm uninstall @p5-wrapper/react
// delete package-lock.json, just in case
npm install p5@1.4.1
```

And then in `package-lock.json` it still reads `"p5": "^1.9.0"`

Any ideas why it would do that? 🤔

**Reply by yevdyko** [reply]
@taliacotton I think we can change requirements for p5 to be less strict. What version of p5 would you like to use? WDYT @jamesrweb ?

**Reply by yevdyko** [reply]
OK, I see and thank you for the feedback. I will try to play with that and come back to you soon

**Reply by jamesrweb** [reply]
We could make it a peer dependency but then we would have no way to ensure the p5 we use is the one the user uses and as mentioned in #318 there are security concerns with lower versioned packages and I want to be sure there's an actual use case which would benefit users enough to counterbalance the trade offs.

**Reply by yevdyko** [reply]
Perhaps this option: "^1.4.1" will help us, so that we can allow the installation of all p5 versions, starting with "1.4.1" and ending with the latest one. WDYT @jamesrweb ?

**Reply by yevdyko** [reply]
More explicit option:

```json
{
  "dependencies": {
    "p5": ">=1.4.1 <2.0.0"
  }
}
```

**Reply by jamesrweb** [reply]
We could but what if someone in future requested 1.3? I guess a range of 1..2 would work also but I think if we are going to allow such a wide range, we shouldn't pay the security and management cost for it, in such a case a peer dependency is probably the best option to offset the risks and costs. WDYT?

**Reply by yevdyko** [reply]
We can only allow this until the bug fix is not released in p5, and adjust it to use the latest version after that

**Reply by jamesrweb** [reply]
Okay, im open to the idea. If we notice security notifications or issues being raised though, for example, then I'd move it to a peer dependency instead.

Would you make a PR for this proposed change and we can try your way first then.

**Reply by jamesrweb** [reply]
But wait, if it's a dependency, won't NPM resolve it to the latest...? I wonder if the user can actually choose or overwrite in such a case. Have you tried this before?

**Reply by yevdyko** [reply]
But the peer dependency probably seems like a more flexible solution, and we don't need to strictly adhere to the preferred version of p5 for users since we only provide a wrapper on top of it. To sum up, I don't mind to move it to peer dependencies

**Reply by yevdyko** [reply]
I think that when a dependency version is specified in a range, it can be overwritten if the user explicitly specifies the required version in their project. If not, the latest version will be installed. I can try this on a demo application

**Reply by yevdyko** [reply]
This is how it looks like in the demo app. There are 2 branches:
1. p5 version as `1.4.1`: https://github.com/yevdyko/react-p5-wrapper-demo/commit/48b20245f7b7f0e49a7695b9229ed8a818e417d3
2. p5 the latest: https://github.com/yevdyko/react-p5-wrapper-demo/commit/36800838fc2f298524a7d67cb0f302f66e4b348f

**Reply by jamesrweb** [reply, marked as answer]
Okay, let's try it this way first and fallback to peer dependencies later if necessary then
